chore: transition the library to microgenerator

plamut opened the pull request:
Closes #41.

**Remaining things:**
 - [x] Add the UPGRADING guide to docs.

plamut commented:
I see potential issues with namespace clashes.

The following imports used with the old generated client do not work anymore:
```py
from google.cloud import bigquery_datatransfer
from google.cloud import bigquery_datatransfer_v1
```

These imports need to be changed to the following:
```py
from google.cloud.bigquery import datatransfer
from google.cloud.bigquery import datatransfer_v1
```

This is all well and good, but if one then installs `google-cloud-bigquery`, these imports _stop working_:
```
$ pip install google-cloud-bigquery
$ python3.8
...
>>> from google.cloud.bigquery import datatransfer
ImportError: cannot import name 'datatransfer' from 'google.cloud.bigquery'
```

Uninstalling `google-cloud-bigquery`  makes the datatransfer import work again.

@busunkim96 Do you maybe know if any other 100% autogenerated libraries experienced the same problem and how they fixed it?

plamut commented on samples/create_scheduled_query.py:
```diff
@@ -68,11 +68,15 @@ def sample_create_transfer_config(project_id, dataset_id, authorization_code="")
             "schedule": "every 24 hours",
         },
-        bigquery_datatransfer_v1.types.TransferConfig(),
+        datatransfer_v1.types.TransferConfig.pb()(),
```
Is there a more clean version of this that doesn't require the user to construct a raw protobuf message?

`google.protobuf.json_format.ParseDict()` currently fails if `TransferCOnfig()` instance is passed in directly (AttributeError: 'DESCRIPTOR').

tswast replied:
I think I used ParseDict because the former client couldn't take a dictionary without failing. Maybe it's fixed now and you don't have to manually call ParseDict?

plamut replied:
Aha, it seems that one can construct a `TransferConfig` instance directly, the test passes:

```py
    transfer_config = datatransfer_v1.types.TransferConfig(
        **{
            "destination_dataset_id": dataset_id,
            "display_name": "Your Scheduled Query Name",
            "data_source_id": "scheduled_query",
            "params": {
                "query": query_string,
                "destination_table_name_template": "your_table_{run_date}",
                "write_disposition": "WRITE_TRUNCATE",
                "partitioning_field": "",
            },
            "schedule": "every 24 hours",
        }
    )
```

(it's even cleaner to provide data as kwargs instead of unpacking a dict, that was just a quick test that I made)

busunkim96 replied:
^ I'd also recommend the kwargs approach, you get hints and such in IDEs which is nicer than a big dictionary. 😄

busunkim96 commented:
@plamut @tswast If you all prefer to retain the old namespacing you can do that via generator args

https://github.com/googleapis/googleapis/blob/b821f320473c8ec05a1c7fb9a496c958b1ab9834/google/devtools/clouderrorreporting/v1beta1/BUILD.bazel#L161

```
opt_args = ["python-gapic-namespace=google.cloud", "python-gapic-name=bigquery_datatransfer"],
```

Not sure where we landed with the other BigQuery libraries though, I assume it's better for them to be consistent?

danoscarmike commented:
Thanks @plamut! I'll defer to @busunkim96 and @tswast on this one.

tswast commented:
> Not sure where we landed with the other BigQuery libraries though, I assume it's better for them to be consistent?

https://github.com/googleapis/python-bigquery-reservation and https://github.com/googleapis/python-bigquery-connection both use the `google.cloud.bigquery` namespace.

plamut commented:
OK, I'll keep the `google.cloud.bigquery` namespace then and document it in the UPGRADING guide accordingly.

plamut commented:
Added the UPGRADING guide and simplified the sample, the PR should be done now.

After approving and merging, I will follow up with a release PR (probably on Monday).

busunkim96 commented on synth.py:
```diff
@@ -22,62 +22,48 @@
 python.py_samples(skip_readmes=True)
 
+# Fix missing async client in datatransfer_v1
+s.replace(
```
The generator isn't adding the async client to `__init__.py`?

plamut replied:
IIRC something tried to import an Async client somewhere, but it did not exist, thus added this import.
